instrumentation/micrometer/micrometer-1.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java

```diff
@@ -0,0 +1,167 @@
+ */
+
+package io.opentelemetry.javaagent.instrumentation.micrometer.v1_5;
+
+import static io.opentelemetry.javaagent.instrumentation.micrometer.v1_5.Bridging.description;
+import static io.opentelemetry.javaagent.instrumentation.micrometer.v1_5.Bridging.toAttributes;
+
+import io.micrometer.core.instrument.AbstractTimer;
+import io.micrometer.core.instrument.Clock;
+import io.micrometer.core.instrument.Measurement;
+import io.micrometer.core.instrument.distribution.DistributionStatisticConfig;
+import io.micrometer.core.instrument.distribution.NoopHistogram;
+import io.micrometer.core.instrument.distribution.TimeWindowMax;
+import io.micrometer.core.instrument.distribution.pause.PauseDetector;
+import io.micrometer.core.instrument.util.TimeUtils;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.metrics.LongHistogram;
+import io.opentelemetry.api.metrics.Meter;
+import java.util.Collections;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+final class OpenTelemetryTimer extends AbstractTimer implements RemovableMeter {
+
+  // TODO: use bound instruments when they're available
+  private final LongHistogram otelHistogram;
+  private final Attributes attributes;
+  private final Measurements measurements;
+
+  volatile boolean removed = false;
+
+  OpenTelemetryTimer(
+      Id id,
+      Clock clock,
+      DistributionStatisticConfig distributionStatisticConfig,
+      PauseDetector pauseDetector,
+      Meter otelMeter) {
+    super(id, clock, distributionStatisticConfig, pauseDetector, TimeUnit.MILLISECONDS, false);
+
+    this.otelHistogram =
+        otelMeter
+            .histogramBuilder(id.getName())
+            .setDescription(description(id))
+            .setUnit("milliseconds")
+            .ofLongs()
+            .build();
+    this.attributes = toAttributes(id.getTags());
+
+    if (isUsingMicrometerHistograms()) {
+      measurements = new MicrometerHistogramMeasurements(clock, distributionStatisticConfig);
+    } else {
+      measurements = NoopMeasurements.INSTANCE;
+    }
+  }
+
+  boolean isUsingMicrometerHistograms() {
+    return histogram != NoopHistogram.INSTANCE;
+  }
+
+  @Override
+  protected void recordNonNegative(long amount, TimeUnit unit) {
+    if (amount >= 0 && !removed) {
+      otelHistogram.record(unit.toMillis(amount), attributes);
+      measurements.record(amount, unit);
+    }
+  }
+
+  @Override
+  public long count() {
+    return measurements.count();
+  }
+
+  @Override
+  public double totalTime(TimeUnit unit) {
+    return measurements.totalTime(unit);
+  }
+
+  @Override
+  public double max(TimeUnit unit) {
+    return measurements.max(unit);
+  }
+
+  @Override
+  public Iterable<Measurement> measure() {
+    // OpenTelemetry metrics bridge does not support reading measurements
+    return Collections.emptyList();
+  }
+
+  @Override
+  public void onRemove() {
+    removed = true;
+  }
+
+  private interface Measurements {
```

Comment:
I struggled to understand the purpose of this interface based on its name alone. Is there something that better captures the abstraction it represents?

Reply:
It quite literally represents the [`Measurement`s](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/Measurement.java) made by the timer, just grouped together. Do you have any suggestion on how to name this?

Reply:
Can't think of a good one.